sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java

```diff
@@ -737,6 +765,9 @@ public PCollectionTuple expand(PCollection<? extends InputT> input) {
       // Use coder registry to determine coders for all StateSpec defined in the fn signature.
       finishSpecifyingStateSpecs(fn, input.getPipeline().getCoderRegistry(), input.getCoder());
 
+      // Ensures that if the fn uses state that the input has a KvCoder
+      validateStateApplicableForInput(fn, input);
```

Comment:
I would consider instead moving the conditional here:

```
if (signature.usesState() || signature.usesTimers()) {
 validateStateApplicableForInput(fn, input);
}
```

Since it makes the condition and things clearer.